simulator/memory/argv_loader/argv_loader.cpp

```diff
@@ -0,0 +1,84 @@
+/**
+ * argv_loader.cpp: loader of arguments to guest memory
+ * @author Rustem Yunusov <yunusov.rr@phystech.edu>
+ * Copyright 2019 MIPT-MIPS iLab project
+ */
+
+#include "argv_loader.h"
+
+#include <cstring>
+#include <infra/exception.h>
+
+template <typename T, Endian endian>
+ArgvLoader<T, endian>::ArgvLoader( const char* const* argv, const char* const* envp)
+        : argc( argv ? count_argc( argv) : 0)
+        , argv( argv)
+        , envp( envp)
+        , offset( 0)
+{}
+
+template <typename T, Endian endian>
+size_t ArgvLoader<T, endian>::load_to( const std::shared_ptr<FuncMemory>& mem, Addr addr)
+{
+    mem->write<T, endian>( narrow_cast<T>( argc), addr + offset);
+    offset += GUEST_WORD_SIZE;
+
+    offset += argc * bytewidth<Addr>; //reserved space for argv[]
+
+    place_nullptr( mem, addr + offset);
+    offset += bytewidth<Addr>;
+
+    if ( envp)
+    {
+        offset += count_argc( envp) * bytewidth<Addr>;
+
+        place_nullptr( mem, addr + offset);
+        offset += bytewidth<Addr>;
+    }
+
+    load_argv_contents( mem, addr);
+
+    if ( envp)
+        load_envp_contents( mem, addr);
+
+    return offset;
+}
+
+template<typename T, Endian endian>
+void ArgvLoader<T, endian>::load_argv_contents( const std::shared_ptr<FuncMemory>& mem, Addr addr)
+{
+    for ( int contents_offset = 0; contents_offset < argc; contents_offset++)
+    {
+        mem->write<Addr, endian>( offset, addr + GUEST_WORD_SIZE + contents_offset * bytewidth<Addr>);
```

Comment:
and here

Reply:
I don't really get it.
'int' content_offset is in host memory.
'bytewidth<Addr> is not bytewidth<T>.